src/nxt_cgroup.c

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) Andrew Clayton
+ * Copyright (C) F5, Inc.
+ */
+
+#include <nxt_main.h>
+#include <nxt_process.h>
+#include <nxt_malloc.h>
+#include <nxt_fs.h>
+#include <nxt_cgroup.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <limits.h>
+
+
+static char *nxt_mk_cgpath(const char *dir);
+
+
+nxt_int_t
+nxt_cgroup_proc_add(const nxt_task_t *task, nxt_process_t *process)
+{
+    int   ret, len;
+    FILE  *fp;
+    char  *cgpath;
+    char  cgprocs[PATH_MAX - 13];   /* - 13 for '/cgroup.procs' */
+
+    if (task->thread->runtime->type != NXT_PROCESS_MAIN
+        || nxt_process_type(process) != NXT_PROCESS_PROTOTYPE
+        || process->isolation.cgroup.path == NULL)
+    {
+        return NXT_OK;
+    }
+
+    cgpath = nxt_mk_cgpath(process->isolation.cgroup.path);
+    if (cgpath == NULL) {
+        return NXT_ERROR;
+    }
+    nxt_fs_mkdir_all((const u_char *) cgpath, 0777);
+
+    snprintf(cgprocs, sizeof(cgprocs), "%s/cgroup.procs", cgpath);
+    fp = fopen(cgprocs, "we");
+    if (fp == NULL) {
+        ret = NXT_ERROR;
+        goto out_free;
+    }
+
+    ret = NXT_OK;
+
+    setvbuf(fp, NULL, _IONBF, 0);
+    len = fprintf(fp, "%d\n", process->pid);
+    if (len < 0) {
+        ret = NXT_ERROR;
+    }
+
+    fclose(fp);
+
+out_free:
+
+    nxt_free(cgpath);
+
+    return ret;
+}
+
+
+void
+nxt_cgroup_cleanup(const nxt_process_t *process)
+{
+    char  *cgroot, *cgpath, *ptr;
+
+    cgroot = nxt_mk_cgpath("");
+    cgpath = nxt_mk_cgpath(process->isolation.cgroup.path);
+    if (cgroot == NULL || cgpath == NULL) {
+        goto out_free;
+    }
+
+    while (*cgpath != '\0' && strcmp(cgroot, cgpath) != 0) {
+        rmdir(cgpath);
+        ptr = strrchr(cgpath, '/');
+        *ptr = '\0';
+    }
+
+out_free:
+
+    nxt_free(cgroot);
+    nxt_free(cgpath);
+}
+
+
+static char *
+nxt_mk_cgpath(const char *dir)
+{
+    char  *path, *buf, *ptr;
+
+    path = buf = ptr = NULL;
+
+    /*
+     * If the path from the config is relative, we need to make
+     * the cgroup path include the main unit processes cgroup. I.e
+     *
+     *   NXT_CGROUP_ROOT/<main process cgroup>/<cgroup path>
+     */
+    if (dir[0] != '/') {
+        int      i;
+        bool     found;
+        FILE     *fp;
+        size_t   len;
+        ssize_t  nread;
+
+        fp = fopen("/proc/self/cgroup", "re");
+        if (fp == NULL) {
+            return NULL;
+        }
+
+        found = false;
+        while ((nread = getline(&buf, &len, fp)) != -1) {
+            if (strncmp(buf, "0::", 3) != 0) {
+                continue;
+            }
+
+            found = true;
+            break;
+        }
+
+        fclose(fp);
+
+        if (found == false) {
+            goto out_free_buf;
+        }
+
+        buf[nread - 1] = '\0';  /* lose the trailing '\n' */
+        ptr = buf;
+        for (i = 0; i < 2; i++) {
```

Comment:
I prefer C99 loop variables (declared in the loop initializer).

Also, I prefer using unsigned loop iterators:

<https://gustedt.wordpress.com/2013/07/15/a-praise-of-size_t-and-other-unsigned-types/>

<https://gustedt.wordpress.com/2013/07/15/a-praise-of-size_t-and-other-unsigned-types/#comment-2953>

<https://git.kernel.org/pub/scm/docs/man-pages/man-pages.git/commit/?id=b42296e4feaffd0ca4e5675fa181d19891f947ce>

Reply:
I don't mind using C99 loop initialisers, they're just not a thing in this code base.

As for signed vs unsigned, I'd say to make the loop counter match the thing your counting against. In this case I can't see it making any difference counting to 2.

Reply:
Yeah, matching is definitely more important than using unsigned.  Nevertheless, I've found that except for a few cases, you normally can just make the thing your counting against unsigned as easily.

In this trivial case it's almost irrelevant, but anyway, I think it makes sense to try to start some consistency here (and the same with the C99 initializers; I very much prefer them).

Reply:
Perhaps requires a patch to the coding style first...

Reply:
> Perhaps requires a patch to the coding style first...

I don't think we even have one :p
We point to the NGINX one, but we already deviate from it in a few cases (see for example how they align declarations of pointers with declarations of variables).

Reply:
@artemkonev

Reply:
This is my only issue with the patch, but it's very minor, and apart from that, it's good, so for `a8b3ade "Isolation: added core cgroup infrastructure."`, also `Reviewed-by: Alejandro Colomar <alx@nginx.com>`.